ddtrace/internal/telemetry/writer.py

```diff
@@ -220,13 +221,35 @@ def __init__(self):
         self._namespace = MetricNamespace()
         self._logs = []  # type: List[Dict[str, Any]]
 
+    def _fork_writer(self):
+        # type: () -> None
+        self.disable()
+        self.reset_queues()
+
+    def enable(self):
+        # type: () -> bool
+        """
+        Enable the instrumentation telemetry collection service. If the service has already been
+        activated before, this method does nothing. Use ``disable`` to turn off the telemetry collection service.
+        """
+        if config._telemetry_metrics_enabled:
+            if self.status == ServiceStatus.RUNNING:
+                return True
+
+            self._enabled = True
+            self.start()
+
+            atexit.register(self.stop)
+            return True
+        return False
+
     def add_log(self, level, message, stack_trace="", tags={}):
         # type: (str, str, str, MetricTagType) -> None
         """
         Queues log. This event is meant to send library logs to Datadog’s backend through the Telemetry intake.
         This will make support cycles easier and ensure we know about potentially silent issues in libraries.
         """
-        if self._enabled:
+        if self.enable():
```

Comment:
Why are we using `self.enable()` rather than checking the flag `self._enabled` here? I'm not sure I agree with using this logic to both enable the writer and check if it's enabled, which I think `self._enabled` should be used for the latter.

Reply:
"The `self._enabled` flag works fine with `app-heartbeat` and `app-started` events because the `TelemetryWriter` is running on the main process. However, I found a problem with `self._enabled` in Gunicorn applications. When Gunicorn spawns a child process, the flag `self._enabled` is still true, but the `TelemetryLogsMetricsWriter` is not running in the child process. As a result, the metrics are collected, and we have a memory leak because any worker sends and flushes this collected data on Gunicorn child processes.

In this PR, `TelemetryLogsMetricsWriter` starts correctly when applications spawn a new child process, and we validate `config._telemetry_metrics_enabled` each time we add a metric to avoid leaks. 

I'm thinking the problem is in the naming because I override the `enabled()` method for `TelemetryLogsMetricsWriter`?

Reply:
Yeah, I like it! I think we could remove first `telemetry_metrics_writer._enabled` and then, in a new PR, remove `telemetry_writer._enabled`. If we change both, I think we are mixing features and this PR could be quite huge, what do you think @mabdinur ?